### PR TITLE
Noam/dee 480 support weak segments in multilabel

### PR DESCRIPTION
**Pull request by noamzbr**
Support for multilabel in weak segments.

### CODE REVIEW

File: deepchecks/nlp/checks/model_evaluation/weak_segments_performance.py [hunk 1]
```diff
@@ -61,8 +61,19 @@ def run_logic(self, context: Context, dataset_kind) -> CheckResult:
                                                          n_top_features=self.n_top_features)
 
         # Decide which scorer and score_per_sample to use in the algorithm run
-        encoded_dataset = self._target_encode_categorical_features_fill_na(features, text_data.label,
-                                                                           cat_features, is_cat_label=True)
+        is_multilabel = text_data.is_multi_label_classification()
+        if is_multilabel:
+            label = TruncatedSVD(1).fit_transform(text_data.label).squeeze()
+            is_cat_label = False
```
Comment (Nadav-Barak) [inline deepchecks/nlp/checks/model_evaluation/weak_segments_performance.py:67]: confusing. remove this variable and use is_multilabel instead

File: deepchecks/nlp/checks/model_evaluation/weak_segments_performance.py [hunk 2]
```diff
@@ -61,8 +61,19 @@ def run_logic(self, context: Context, dataset_kind) -> CheckResult:
                                                          n_top_features=self.n_top_features)
 
         # Decide which scorer and score_per_sample to use in the algorithm run
-        encoded_dataset = self._target_encode_categorical_features_fill_na(features, text_data.label,
-                                                                           cat_features, is_cat_label=True)
+        is_multilabel = text_data.is_multi_label_classification()
+        if is_multilabel:
+            label = TruncatedSVD(1).fit_transform(text_data.label).squeeze()
```
Comment (Nadav-Barak) [inline deepchecks/nlp/checks/model_evaluation/weak_segments_performance.py:66]: add comment what this do

File: deepchecks/nlp/checks/model_evaluation/weak_segments_performance.py [hunk 3]
```diff
@@ -61,8 +61,19 @@ def run_logic(self, context: Context, dataset_kind) -> CheckResult:
...
+            is_cat_label = False
+        else:
+            label = text_data.label
+            is_cat_label = True
+        encoded_dataset = self._target_encode_categorical_features_fill_na(features, label,
+                                                                           cat_features, is_cat_label=is_cat_label)
```
Comment (Nadav-Barak) [inline deepchecks/nlp/checks/model_evaluation/weak_segments_performance.py:72]: in NLP is_cat_label always true

File: deepchecks/nlp/checks/model_evaluation/weak_segments_performance.py [hunk 4]
```diff
@@ -61,8 +61,19 @@ def run_logic(self, context: Context, dataset_kind) -> CheckResult:
...
+        # Replacing the label with the original label for multilabel
+        if is_multilabel:
+            listed_label = [list(x) for x in text_data.label]
+            encoded_dataset._data[encoded_dataset.label_name] = listed_label  # pylint: disable=protected-access
```
Comment (Nadav-Barak) [inline deepchecks/nlp/checks/model_evaluation/weak_segments_performance.py:76]: thats ugly and unneeded, just dont use encoded_dataset.label_col

File: deepchecks/tabular/context.py [hunk 5]
```diff
@@ -63,32 +63,41 @@ class _DummyModel:
...
+        test: Dataset,
+        y_proba_test: t.Optional[np.ndarray] = None,
+        y_pred_test: t.Optional[np.ndarray] = None,
+        train: t.Optional[Dataset] = None,
+        y_pred_train: t.Optional[np.ndarray] = None,
+        y_proba_train: t.Optional[np.ndarray] = None,
+        validate_data_on_predict: bool = True,
+        model_classes: t.Optional[t.List[t.Any]] = None
+    ):
+        # TODO:
```
Comment (Nadav-Barak) [inline deepchecks/tabular/context.py:77]: Cause its not validated in the run method i guess

File: deepchecks/tabular/context.py [hunk 6]
```diff
@@ -63,32 +63,41 @@ class _DummyModel:
...
+        # why do we check it then?
         if train is not None and test is not None:
             # check if datasets have same indexes
-            if set(train.data.index) & set(test.data.index):
-                train.data.index = map(lambda x: f'train-{x}', list(train.data.index))
-                test.data.index = map(lambda x: f'test-{x}', list(test.data.index))
-                get_logger().warning('train and test datasets have common index - adding "train"/"test"'
-                                     ' prefixes. To avoid that provide datasets with no common indexes '
-                                     'or pass the model object instead of the predictions.')
+            train_index = train.data.index
```
Comment (Nadav-Barak) [inline deepchecks/tabular/context.py:82]: redundant, no need to create those variables - they have no further use

File: deepchecks/tabular/metric_utils/scorers.py [hunk 7]
```diff
@@ -267,6 +268,10 @@ def predict(self, data: pd.DataFrame) -> np.ndarray:
                     predictions = transfer_func(predictions)
                 # In case of multiclass with single label, convert into multi-label
                 elif self.model_classes:
+                    # if multilabel convert from numpy array of lists to 2d numpy array
+                    if len(predictions) != 0:
```
Comment (Nadav-Barak) [inline deepchecks/tabular/metric_utils/scorers.py:272]: can merge the if

File: deepchecks/utils/single_sample_metrics.py [hunk 8]
```diff
@@ -24,26 +24,30 @@ def calculate_neg_mse_per_sample(labels, predictions, index=None) -> pd.Series:
     return pd.Series([-(y - y_pred) ** 2 for y, y_pred in zip(labels, predictions)], index=index)
 
 
-def calculate_neg_cross_entropy_per_sample(labels, probas: np.ndarray, model_classes: Optional[List] = None,
-                                           index=None, eps=1e-15) -> pd.Series:
+def calculate_neg_cross_entropy_per_sample(labels, probas: np.ndarray,
```
Comment (Nadav-Barak) [inline deepchecks/utils/single_sample_metrics.py:27]: add tests for multi-label (can copy structure from my PR)
Most important to verify running it on the multilabel format for regular classification return the same result

**Review by Nadav-Barak — APPROVED** [review]

File: deepchecks/tabular/context.py (hunk 6 above)
Comment (noamzbr) [reply deepchecks/tabular/context.py:82]: used twice so I'll keep them
